src/layer.rs

```diff
@@ -1,102 +1,732 @@
-    pub layout: Option<Layout>,
+    #[serde(rename = "type")]
+    pub type_: String,
     #[serde(skip_serializing_if = "Option::is_none")]
     pub maxzoom: Option<f64>,
     #[serde(skip_serializing_if = "Option::is_none")]
     pub minzoom: Option<f64>,
     #[serde(skip_serializing_if = "Option::is_none")]
-    pub paint: Option<Paint>,
+    pub visibility: Option<Visibility>,
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase", tag = "type")]
+pub enum Layer {
+    Custom(CustomLayer),
+    Fill(FillLayer),
+    Background(BackgroundLayer),
+    Line(LineLayer),
+    Symbol(SymbolLayer),
+    Circle(CircleLayer),
+}
+
+pub trait IntoLayer {
+    fn into_layer(self) -> Layer;
+}
+
+#[derive(Debug, Clone, Default, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct LayerBase {
+    pub id: String,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub maxzoom: Option<f64>,
     #[serde(skip_serializing_if = "Option::is_none")]
-    pub rendering_mode: Option<String>,
+    pub minzoom: Option<f64>,
     pub source: String,
     #[serde(skip_serializing_if = "Option::is_none")]
+    pub filter: Option<Expression<()>>,
+    #[serde(skip_serializing_if = "Option::is_none")]
     #[serde(rename = "source-layer")]
     pub source_layer: Option<String>,
-    pub r#type: String,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub slot: Option<String>,
 }
 
-impl Layer {
-    pub fn new(
-        id: impl Into<String>,
-        r#type: impl Into<String>,
-        source: impl Into<String>,
-    ) -> Layer {
-        Layer {
+impl LayerBase {
+    pub fn new(id: impl Into<String>, source: impl Into<String>) -> LayerBase {
+        LayerBase {
             id: id.into(),
-            layout: None,
             maxzoom: None,
             minzoom: None,
-            paint: None,
-            rendering_mode: None,
             source: source.into(),
+            filter: None,
             source_layer: None,
-            r#type: r#type.into(),
+            slot: None,
         }
     }
 }
 
-/// Layout property can be either value in String or Number e.g. 0.25
-/// or tuple of 2 elements, ("get", "icon").
-#[derive(Debug, Clone, Serialize)]
-#[serde(untagged)]
-pub enum LayoutProperty {
-    String(String),
-    Number(f64),
-    FeatureProperty((String, String)),
+#[derive(Debug, Clone, Default, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct CustomLayer {
+    #[serde(flatten)]
+    pub inner: LayerBase,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub rendering_mode: Option<String>,
+    #[serde(with = "serde_wasm_bindgen::preserve")]
+    pub on_add: js_sys::Function,
+    #[serde(with = "serde_wasm_bindgen::preserve")]
+    pub on_remove: js_sys::Function,
+    #[serde(with = "serde_wasm_bindgen::preserve")]
+    pub prerender: js_sys::Function,
+    #[serde(with = "serde_wasm_bindgen::preserve")]
+    pub render: js_sys::Function,
```

Comment:
I think there might be other ways of moving these closures through the wasm boundary but this was the first one I figured out. If you know of a better approach I'm all ears. This is my first real foray into wasm-bindgen land so idk. I'm just winging it.

Reply:
hmm I don't know the best practice ..